### PR TITLE
[react] Fix return value of convenience overload

### PR DESCRIPTION
**Pull request by LinusU**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://reactjs.org/docs/hooks-reference.html#useref
- [n/a] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

-----

I'm open to discussing this change, but the previous behaviour didn't really make any sense to me. This is the note that was removed:

```text
* Usage note: if you need the result of useRef to be directly mutable, include `| null` in the type
* of the generic argument.
```

This overload happens when doing e.g. this:

```ts
const ref = useRef<number>(null)
```

The returned `ref` is now immutable, but I don't see a use case for that at all. If it actually was immutable when starting with the `null` value, then it would never be mutated to hold anything other than `null` and thus it would be useless as each usage of it could just be replaced with `null`.

-----

Another change was to make it so that `useRef()` and `useRef(undefined)` works exactly the same. React doesn't do any arity-checks, so the two calls are exactly identical. They both create a ref-object with the initial value of `undefined`.

### CODE REVIEW

**Comment by ferdaber** [comment]
The reasoning behind the readonly overload is for cases where the ref is only used as a ref attribute. With the advent of hooks, there are two intents of `useRef()`, one for React-owned refs, and one for userland-owned refs:

### React-owned ref
```tsx
function Component() {
  // this is the readonly overload, this should *not* be mutated within the render function. React owns this ref and manages
  // its pointer value
  const containerRef = useRef<HTMLDivElement>(null) ;
  return <div ref={containerRef} />;
}
```

### Userland-owned ref
```tsx
function Component() {
  // this is the mutable overload, the intent being that it will never be passed into a ref attribute
  // you control this ref and when to update it
  const someNonReactiveCounter = useRef(0);
  
  useEffect(() => {
     if (someCondition) {
       someNonReactiveCounter.current++;
     }
  }, [...stuff]);
  
  return <div />;
}
```

As far as the `undefined` overload change, the reason isn't because of runtime arity check, it's a convenience factor so that the user doesn't have to check for undefined each time, with your change, the type would always be `T | undefined`:

function Component() {
  // this is the mutable overload, the intent being that it will never be passed into a ref attribute
  // you control this ref and when to update it
  const someNonReactiveCounter = useRef(0);
  
  useEffect(() => {
     if (someCondition) {
	   // this fails because TS thinks it could be undefined, even though you never actually intend to ever unset it
       someNonReactiveCounter.current++;
     }
  }, [...stuff]);
  
  return <div />;
}

**Comment by eps1lon** [comment]
>   // the type is MutableRef<number | undefined> instead of MutableRef<number>
>  const someNonReactiveCounter = useRef(0);

Do we have a test for that? Because I don't think this should be `T | undefined`. This will inconvenience every usage of `useRef` for the "previous props pattern".

**Comment by LinusU** [comment]
@ferdaber Thanks for taking the time to explain!

> The reasoning behind the readonly overload is for cases where the ref is only used as a ref attribute. With the advent of hooks, there are two intents of `useRef()`, one for React-owned refs, and one for userland-owned refs:
> 
> ### React-owned ref
> ```tsx
> function Component() {
>   // this is the readonly overload, this should *not* be mutated within the render function. React owns this ref and manages
>   // its pointer value
>   const containerRef = useRef<HTMLDivElement>(null) ;
>   return <div ref={containerRef} />;
> }
> ```
> 
> ### Userland-owned ref
> ```tsx
> function Component() {
>   // this is the mutable overload, the intent being that it will never be passed into a ref attribute
>   // you control this ref and when to update it
>   const someNonReactiveCounter = useRef(0);
>   
>   useEffect(() => {
>      if (someCondition) {
>        someNonReactiveCounter.current++;
>      }
>   }, [...stuff]);
>   
>   return <div />;
> }
> ```

Hmm, this is neat when only using built-ins, but does this mean that the recommended way to build your own components that take a `ref` is to accept "non-mutable" ref object, and than force cast it to a mutable one in order to actually assign something?

I _think_ that this _feels_ a bit like we've have added more things in the types here than the original library had. E.g. if React was written in TypeScript, would it accept non-mutable ref objects and force cast them to mutable in their code? That sounds a bit strange to me...

What it the benefit of having it not be mutable from user space? I cannot think of any pitfalls that this would prevent, since you could only assign e.g. a `HTMLDivElement` to it, which you probably won't do unless you intended to?

-----

> As far as the `undefined` overload change, the reason isn't because of runtime arity check, it's a convenience factor so that the user doesn't have to check for undefined each time, with your change, the type would always be `T | undefined`:
> 
> ```tsx
> function Component() {
>   // the type is MutableRef<number | undefined> instead of MutableRef<number>
>   const someNonReactiveCounter = useRef(0);
>   
>   useEffect(() => {
>      if (someCondition) {
> 	   // this fails because TS thinks it could be undefined, even though you never actually intend to ever unset it
>        someNonReactiveCounter.current++;
>      }
>   }, [...stuff]);
>   
>   return <div />;
> }
> ```

I don't think that this is the case? I've included tests that makes sure that it doing `useRef(0)` returns a `MutableRef<number>` and not `MutableRef<number | undefined>`. That overload should only kick in when doing `useRef<number>()` or `useRef<number>(undefined)`.

**Comment by ferdaber** [comment]
It depends on the workflow, if you build a component that takes a ref, with hooks you'd likely use either `useImperativeHandle` or passing it a ref attribute to a host element, both of which also require `forwardRef`. Both cases already take readonly refs as arguments:

```ts
const MyComponentWithAMethod = forwardRef(
  (props: {}, ref: Ref<{ doSomething(): void }>) => {
    useImperativeHandle(ref, () => {
      return { doSomething() { console.log('hello world!'); } }
    });
  })
)

const MyComponent() {
  const customRef = useRef<{ doSomething(): void }>(null);
  const hostRef = useRef<HTMLDivElement>(null);
  return (
    <>
      <MyComponentWithAMethod ref={customRef} /> {/*this works fine*/}
      <div ref={hostRef} /> {/*this works fine*/}
    </>
  );
}
```

**Comment by LinusU** [comment]
Ahh, I see, I realise that why I ran into this initially was because I was doing `const foo = useRef<number>(null)` and then tried to assign `foo.current`, only I was passing the ref to another component which made it even more unclear what went wrong. I now understand the background to how this API is typed.

So the intention is that `useRef(null)` should _only_ be used for references to other components, and `useRef()` and `useRef(/* anything but null */)` should _only_ be used when not referencing components?

-----

I still feel that this is a bit surprising when looked from a perspective of someone being very used to TypeScript, and not too much with the React types.

e.g. we can now construct a function that takes `{ readonly current: T }` as an argument, but that does in fact mutate `current`.  While I can see the background that lead to this, I'm still not sure what the _benefit_ is of doing it this way.

> What it the benefit of having it not be mutable from user space? I cannot think of any pitfalls that this would prevent, since you could only assign e.g. a `HTMLDivElement` to it, which you probably won't do unless you intended to?

If there is an answer to this question ☝️ it would help me understand here ☺️ 

Thanks!

**Comment by ferdaber** [comment]
If someone were to manually set the `.current` value of a ref object that is *also* passed in as a ref parameter, there is a strong chance that the value will be overwritten (think shared memory semantics) by the React runtime, so what they get back will not be what they expect. That being said, this is not a pitfall that is commonly-occurring as you said.

The addition of the read-only overloads predates my time as a maintainer, though I am familiar with the justifications behind it. I'm not particularly opinionated on keeping it vs. removing it, but if we are heading in the direction of less type safety, the `Ref` type itself will need to be changed as well.

**Comment by LinusU** [comment]
> If someone were to manually set the `.current` value of a ref object that is _also_ passed in as a ref parameter, there is a strong chance that the value will be overwritten (think shared memory semantics) by the React runtime, so what they get back will not be what they expect.

Potentially, but I feel that this is abusing the type system a bit since we are passing an immutable object to a function that will in fact mutate it. So coming from a TypeScript background but being new to React I think that this is more confusing then helping...

> I'm not particularly opinionated on keeping it vs. removing it, but if we are heading in the direction of less type safety, [...]

I don't believe that this change is "heading in the direction of less type safety". In fact, I think that this is heading towards more sound types which I believe is a good thing and makes it more "type safe".

Included in [the list of common misstakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes) there is an entry that the input type should be read-only if "a function does not write to its parameters". Here we have a function with a readonly type that _does_ modify that parameter.

If you feel like it would be a good idea, we could ping some DT maintainers to look at this as well?

> [...] the `Ref` type itself will need to be changed as well.

Ah yes, it probably doesn't need to be split up between `MutableRefObject` and `RefObject` at all now. Although that could be nice so that a component can indicate wether it takes a reference just for reading, or also for mutating (this was in fact what I was doing when I stumbled upon this in the first place). But I think that it would be nice to use the names `RefObject` and `ReadonlyRefObject` to match the built in TypeScript types naming convention...

Happy that we are having a productive discussion! 🙌

**Comment by gabritto** [comment]
It seems like this still needs further discussions by people from the React community. I think before we can merge this, we need to make sure this change is not too disruptive for users of React, e.g. that it fits the way `useRef` is used and should be used. If you think this is the case, please get other owners/community members to review this, and provide evidence that the changes won't break a lot of people using this, or the cost/benefit of this change. Thank you for your patience.

**Comment by LinusU** [comment]
@gabritto 

> If you think this is the case, please get other owners/community members to review this, [...]

I'm not sure how I would go about doing this more than pining all the maintainers of this package here which have already been done multiple times?

> [...] and provide evidence that the changes won't break a lot of people using this, or the cost/benefit of this change.

I think that the cost/benefit have been debated earlier in this thread. Is there anything that you feel is missing? Happy to add more information!

**Comment by eps1lon** [comment]
@LinusU  I think https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56934#issuecomment-957976407 is unadressed.

**Comment by LinusU** [comment]
@eps1lon missed to answer that, but it's already covered in the test ☺️ 

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56934/files#diff-8036c55d34934cdbf8a09e220f6595f5d994213f44699b0c012f6da99d193a33R127-R128

File: types/react/test/hooks.tsx
Comment (eps1lon) [inline types/react/test/hooks.tsx:161]: Are we sure this is what we want? For literal values I can see the point. But what if we have e.g. `React.useRef<number>(props.someNullableNumber)`? I guess type-checking still works when accessing this ref but it might be passed to something that's not type-safe (e.g. `doSomethingWithANonNullableNumber(ref.current)`) which means the code is no longer type-safe. Before, we could at least have a stop-gap when instantiating the ref.

What happens for `React.useRef<number>('not a number')`? I would expect an error instead of `MutableRefObject<number | string>`. It doesn't look like we have a test for that.

**Review by eps1lon — CHANGES_REQUESTED** [review]

File: types/react/test/hooks.tsx
Comment (LinusU) [reply types/react/test/hooks.tsx:161]: > Are we sure this is what we want? For literal values I can see the point. But what if we have e.g. React.useRef<number>(props.someNullableNumber)?

I'd be open to letting `useRef<number>(foo as number | null)` give a type error.

In fact, the thing I'm most interested in is `useRef<number>()` returning `MutableRefObject<number | undefined>`.

> I guess type-checking still works when accessing this ref but it might be passed to something that's not type-safe (e.g. doSomethingWithANonNullableNumber(ref.current)) which means the code is no longer type-safe. Before, we could at least have a stop-gap when instantiating the ref.

In this case `ref.current` will be typed as `number | null` so the type checking would work as expected.

> What happens for React.useRef<number>('not a number')? I would expect an error instead of MutableRefObject<number | string>. It doesn't look like we have a test for that.

Pushed a commit that adds these tests:

Screenshot 2021-11-30 at 18 35 15

Comment (eps1lon) [reply types/react/test/hooks.tsx:161]: > In fact, the thing I'm most interested in is useRef<number>() returning MutableRefObject<number | undefined>.

I'm in favor of this behavior considering this is equivalent to a literal `undefined`. There's still room for a mistake but that mistake is just a few code columns away so I think that we don't necessarily need to catch it.

 > In this case ref.current will be typed as number | null so the type checking would work as expected.

Intended by you maybe but the resulting behavior is not necessarily expected. Consider you pass this `ref.current` to a function that's loosely typed e.g. `handleRef(value: any)`. Before, you could be sure that `useRef` prevents these mistakes. Your approach assumes that the whole codebase is type-checked which is rarely the case (especially with library types).

**Review by amcasey — CHANGES_REQUESTED** [review]
It sounds like @eps1lon still has some concerns.

File: types/react/test/hooks.tsx
Comment (eps1lon) [reply types/react/test/hooks.tsx:161]: Maybe this wasn't very clear: 
To me `const numberRef = useRef<number>()` might be a mistake (forgot to pass the initial value). That mistake is caught when trying to access `numberRef.current` and expecting a number. However, when `numberRef` is passed to a part of the codebase that is untyped e.g. `adder(numberRef.current, 2)` where `adder(a: any, b: any): number` then we'll never catch this mistake at the type level.

The unfortunate part is that we'll never know which scenario is more likely. But considering you can already get the desired behavior with `useRef<number | undefined>` I'm inclined to not merge this proposal.

My underlying principal is that I try to publish types that are strict by default. The library user can always loosen them up but rarely make them stricter. That's not matching the TypeScript mentality (pragmatically typed language) but I don't find this proposal convincing with regards to pragmatism.

Though the real problem is that we're just two people discussing this. I wouldn't block this proposal if another maintainer would vote for merging this proposal.

**Pull request closed without merging**
